File: src/utils/qty.ts
const UNIT_PATTERN = /(шт|штук|pcs|pc|м\.?|метр|kg|кг|уп\.?|компл\.?)/i;

export interface ParsedQty {
  qty: number | null;
  unit: string | null;
  qtyRaw: string | null;
}

export function parseQty(input: string): ParsedQty {
  const line = input.replace(/\u00A0/g, ' ');
  const numberPattern = '(\\d{1,3}(?:[\\s.,]\\d{3})+|\\d+(?:[.,]\\d+)?)';
  const withUnitMatches = Array.from(
    line.matchAll(new RegExp(`(?<![A-Za-zА-Яа-я0-9.,])${numberPattern}\\s*${UNIT_PATTERN.source}`, 'gi')),
  );
  const numericMatches = Array.from(
    line.matchAll(new RegExp(`(?<![A-Za-zА-Яа-я0-9.,])${numberPattern}(?![A-Za-zА-Яа-я0-9.,])`, 'g')),
  );
  const qtyMatch = withUnitMatches.length
    ? withUnitMatches[withUnitMatches.length - 1]
    : numericMatches.length
      ? numericMatches[numericMatches.length - 1]
      : null;

  let qty: number | null = null;
  let qtyRaw: string | null = null;

  if (qtyMatch) {
    qtyRaw = qtyMatch[0]?.trim() ?? null;
    const primary = (qtyMatch[1] ?? qtyRaw ?? '').trim();
    const normalized = normalizeNumericToken(primary);
    const parsed = Number(normalized);
    qty = Number.isFinite(parsed) ? parsed : null;
  }

  const unitMatch = line.match(UNIT_PATTERN);
  const unit = unitMatch ? normalizeUnit(unitMatch[1]) : null;

  return { qty, unit, qtyRaw };
}

export function normalizeUnit(unit: string): string {
  const u = unit.toLowerCase();
  if (['шт', 'штук', 'pcs', 'pc'].includes(u)) {
    return 'шт';
  }
  if (['м', 'м.', 'метр'].includes(u)) {
    return 'м';
  }
  if (['kg', 'кг'].includes(u)) {
    return 'кг';
  }
  if (['уп', 'уп.'].includes(u)) {
    return 'уп';
  }
  return u;
}

function normalizeNumericToken(token: string): string {
  const compact = token.replace(/\s+/g, '');

  if (/^\d{1,3}(?:\.\d{3})+$/.test(compact)) {
    return compact.replace(/\./g, '');
  }

  if (/^\d{1,3}(?:,\d{3})+$/.test(compact)) {
    return compact.replace(/,/g, '');
  }

  if (compact.includes(',') && !compact.includes('.')) {
    return compact.replace(',', '.');
  }

  return compact;
}

File: internal/util/qty.go
package util

import (
	"regexp"
	"strconv"
	"strings"
)

var (
	unitPattern   = regexp.MustCompile(`(?i)(шт|штук|pcs|pc|м\.?|метр|kg|кг|уп\.?|компл\.?)`)
	numberPattern = regexp.MustCompile(`(?i)(?:^|[^0-9.,])(\d{1,3}(?:[\s.,]\d{3})+|\d+(?:[.,]\d+)?)`)
)

type ParsedQty struct {
	Qty    *float64
	Unit   *string
	QtyRaw *string
}

func ParseQty(input string) ParsedQty {
	line := strings.ReplaceAll(input, "\u00A0", " ")

	qtyRaw := ""
	qtyToken := ""

	withUnit := regexp.MustCompile(`(?i)(?:^|[^0-9.,])(\d{1,3}(?:[\s.,]\d{3})+|\d+(?:[.,]\d+)?)\s*(шт|штук|pcs|pc|м\.?|метр|kg|кг|уп\.?|компл\.?)`)
	wm := withUnit.FindAllStringSubmatch(line, -1)
	if len(wm) > 0 {
		last := wm[len(wm)-1]
		qtyRaw = strings.TrimSpace(last[1] + " " + last[2])
		qtyToken = strings.TrimSpace(last[1])
	} else {
		nm := numberPattern.FindAllStringSubmatch(line, -1)
		if len(nm) > 0 {
			last := nm[len(nm)-1]
			qtyRaw = strings.TrimSpace(last[1])
			qtyToken = strings.TrimSpace(last[1])
		}
	}

	var qtyPtr *float64
	if qtyToken != "" {
		norm := normalizeNumericToken(qtyToken)
		if parsed, err := strconv.ParseFloat(norm, 64); err == nil {
			qtyPtr = FloatPtr(parsed)
		}
	}

	var unitPtr *string
	if um := unitPattern.FindStringSubmatch(line); len(um) > 1 {
		u := normalizeUnit(um[1])
		unitPtr = &u
	}

	var qtyRawPtr *string
	if qtyRaw != "" {
		qtyRawPtr = &qtyRaw
	}

	return ParsedQty{Qty: qtyPtr, Unit: unitPtr, QtyRaw: qtyRawPtr}
}

func normalizeUnit(unit string) string {
	u := strings.ToLower(strings.TrimSpace(unit))
	switch u {
	case "шт", "штук", "pcs", "pc":
		return "шт"
	case "м", "м.", "метр":
		return "м"
	case "kg", "кг":
		return "кг"
	case "уп", "уп.":
		return "уп"
	default:
		return u
	}
}

func normalizeNumericToken(token string) string {
	compact := strings.ReplaceAll(token, " ", "")
	if regexp.MustCompile(`^\d{1,3}(?:\.\d{3})+$`).MatchString(compact) {
		return strings.ReplaceAll(compact, ".", "")
	}
	if regexp.MustCompile(`^\d{1,3}(?:,\d{3})+$`).MatchString(compact) {
		return strings.ReplaceAll(compact, ",", "")
	}
	if strings.Contains(compact, ",") && !strings.Contains(compact, ".") {
		return strings.ReplaceAll(compact, ",", ".")
	}
	return compact
}
